### ISSUE TITLE
Error: EISDIR: illegal operation on a directory, open './assets/css'

### ISSUE THREAD

**Issue by mavanmanen** [gha-utilities/sass-build#14]
It seems I do not have access to the output folder, I have confirmed it exists with an `ls` command, below if the error output and my workflow file

```
Run gha-utilities/sass-build@v0.1.9
/home/runner/work/_actions/gha-utilities/sass-build/v0.1.9/index.js:127
    throw err;
    ^

[Error: EISDIR: illegal operation on a directory, open './assets/css'] {
  errno: -21,
  code: 'EISDIR',
  syscall: 'open',
  path: './assets/css'
}
```

```
name: CI

on:
  push:
    branches: 
      - develop

jobs:
  build_scss:
    name: Build SCSS
    runs-on: ubuntu-latest
     
    steps:
      - name: Checkout source Git branch
        uses: actions/checkout@v2
        with:
          ref: develop
          
      - name: List files
        run: ls -a
       
      - name: Compile CSS from SCSS files
        uses: gha-utilities/sass-build@v0.1.9
        with:
          source: ./_scss/main.scss
          destination: ./assets/css
           
      - name: Checkout destination Git branch
        uses: actions/checkout@v2
        with:
          ref: master
          fetch-depth: 1
         
      - name: Add and Commit changes to master branch
        run: |
         git config --local user.email 'action@github.com'
         git config --local user.name 'GitHub Action'
         git add assets/css/*
         git commit -m 'Updates compiled CSS files'
          
      - name: Push changes
        uses: ad-m/github-push-action@master
        with:
         github_token: ${{ secrets.GITHUB_TOKEN }}
         branch: master
```

**Comment by S0AndS0** [comment]
Thanks for opening this issue!


## TLDR


Please try assigning `destination` to a file path instead of a directory, eg...


**Snip `.github/workflows/build_css.yml`**


```YAML
# ...
      - name: Compile CSS from SCSS files
        uses: gha-utilities/sass-build@v0.1.9
        with:
          source: ./_scss/main.scss
          destination: ./assets/css/main.css
# ...
```


------


**Full `.github/workflows/build_css.yml`**


```YAML
jobs:
  build_scss:
    name: Build SCSS
    runs-on: ubuntu-latest
     
    steps:
      - name: Checkout source Git branch
        uses: actions/checkout@v2
        with:
          ref: develop
          
      - name: List files
        run: ls -a ./assets/css || mkdir -vp ./assets/css
       
      - name: Compile CSS from SCSS files
        uses: gha-utilities/sass-build@v0.1.9
        with:
          source: ./_scss/main.scss
          destination: ./assets/css/main.css
           
      - name: Checkout destination Git branch
        uses: actions/checkout@v2
        with:
          ref: master
          fetch-depth: 1
         
      - name: Add and Commit changes to master branch
        run: |
         git config --local user.email 'action@github.com'
         git config --local user.name 'GitHub Action'
         git add assets/css/*
         git commit -m 'Updates compiled CSS files'
          
      - name: Push changes
        uses: ad-m/github-push-action@master
        with:
         github_token: ${{ secrets.GITHUB_TOKEN }}
         branch: master
```


___


Because I can reproduce this error by attempting to write to a directory...


## Test setup


```Bash
cd ~/git/hub/gha-utilities/sass-build

mkdir -vp .tests/{lib,out}
```


### Test files


**`.tests/lib/something.scss`**


```SCSS
@mixin onething {
  @debug 'mixing something has been called';
}

.another-thing {
  width: auto;
}
```


**`.tests/main.scss`**


```SCSS
@use 'lib/something';

.example-class {
  width: auto;
  height: auto;
  @include something.onething;
}
```


### Test command one


```Bash
INPUT_SOURCE=.tests/main.scss\
  INPUT_DESTINATION=.tests/out\
  npm run test
```


### Error output one


```
> sass-build@0.1.9 test ~/git/hub/gha-utilities/sass-build
> node index.js

.tests/lib/something.scss:2 DEBUG: mixing something has been called
~/git/hub/gha-utilities/sass-build/index.js:127
    throw err;
    ^

[Error: EISDIR: illegal operation on a directory, open '.tests/out'] {
  errno: -21,
  code: 'EISDIR',
  syscall: 'open',
  path: '.tests/out'
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! sass-build@0.1.9 test: `node index.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the sass-build@0.1.9 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     ~/.npm/_logs/2020-02-28___-debug.log
```


------


It's also possible to receive similar (though not identical) errors when writing to a non-existent directory path...


### Test command two


```Bash
INPUT_SOURCE=.tests/main.scss\
  INPUT_DESTINATION=.tests/elsewhere/main.css\
  npm run test
```


### Error output two


```
> sass-build@0.1.9 test ~/git/hub/gha-utilities/sass-build
> node index.js

.tests/lib/something.scss:2 DEBUG: mixing something has been called
~/git/hub/gha-utilities/sass-build/index.js:127
    throw err;
    ^

[Error: ENOENT: no such file or directory, open '.tests/elsewhere/main.css'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '.tests/elsewhere/main.css'
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! sass-build@0.1.9 test: `node index.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the sass-build@0.1.9 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     ~/.npm/_logs/2020-02-28___-debug.log
```


------


Compilation is successful when writing to a file path...


### Test command three


```Bash
INPUT_SOURCE=.tests/main.scss\
  INPUT_DESTINATION=.tests/out/main.css\
  npm run test
```


### Output three


```
> sass-build@0.1.9 test ~/git/hub/gha-utilities/sass-build
> node index.js

.tests/lib/something.scss:2 DEBUG: mixing something has been called
Wrote file -> .tests/out/main.css
```


___


## Notes


For _`Test command one`_ and _`Error output one`_, error state is caused by trying to write to a directory like a file, _"illegal operation on a directory"_

For _`Test command two`_ and _`Error output two`_, error state is caused by trying to write to a file with missing directory(ies), _"no such file or directory"_

Another way to reach _`Error output two`_ is when Git refuses to push/track an empty directory, in such cases it's recommended to either place a `.gitkeep` file within the last directory of such paths, eg...


```Bash
mkdir ./some/where

touch ./some/where/.gitkeep

git add ./some/where/.gitkeep

git commit -m 'Tracks empty directory for _CI-something_ to write to'
```


... or ensure that the directory the path exists prior to writing files within...


**Snip `.github/workflows/build_css.yml`**


```YAML
      - name: List files
        run: ls -a ./assets/css || mkdir -vp ./assets/css
```


... generally _`Error output two`_ is only reachable during branch _boot-strapping_; or in other-words, the first time that CI/CD builds output. Because after successful initialization, Continuous Integration/Deployment steps will overwrite a previously tracked file.

**Comment by S0AndS0** [comment]
I'll be closing this Issue in the next day or two unless I read-back from ya that v0.2.0 hasn't resolved errors.

Thanks again for opening this Issue, and feel free to open new Issues for new bugs or re-open this one if some regression causes these errors to pop again
